pkg/storegateway/indexheader/lazy_binary_reader.go

```diff
@@ -231,124 +269,169 @@ func (r *LazyBinaryReader) EagerLoad(ctx context.Context) {
 // Returns the reader, wait group that should be used to signal that usage of reader is finished, and an error on failure.
 // Must be called without lock.
 func (r *LazyBinaryReader) getOrLoadReader(ctx context.Context) (Reader, *sync.WaitGroup, error) {
-	r.readerMx.RLock()
-	defer r.readerMx.RUnlock()
-
-	// Nothing to do if we already tried loading it.
-	if r.reader != nil {
-		r.usedAt.Store(time.Now().UnixNano())
-		r.readerInUse.Add(1)
-
-		return r.reader, &r.readerInUse, nil
-	}
-	if r.readerErr != nil {
-		return nil, nil, r.readerErr
+	loadedR := r.getReader(ctx)
+	if loadedR.reader != nil {
+		return loadedR.reader, loadedR.inUse, nil
 	}
 
-	// Release the read lock, so that loadReader can take write lock. Take the read lock again once done.
-	r.readerMx.RUnlock()
-	err := r.loadReader()
-	// Re-acquire read lock.
-	r.readerMx.RLock()
+	promise := r.triggerLoadReader(ctx)
+	loadedR = r.waitLoadedReader(ctx, promise)
```

Comment:
It seems that every time we try to `getReader`, we'll load one if there isn't one yet. 

So why is this logic here, instead of having just a single channel to get the reader, and let the `controlLoop` take care of loading one if needed?

Reply:
oh, good point, this makes it much cleaner. I will refactor